### PR TITLE
If allowed to submit, auto-approve and sign experiments (bug 1220715)

### PR DESCRIPTION
**Pull request by magopian**
Fixes [bug 1220715](https://bugzilla.mozilla.org/show_bug.cgi?id=1220715)

This is based on PR #889, only the last commit is relevant to this bug.

### CODE REVIEW

File: apps/files/tests/test_utils_.py [hunk 1]
```diff
@@ -251,8 +251,8 @@ def test_instanciate_without_data(self):
         with NamedTemporaryFile() as file_:
             file_.write(json.dumps(data))
             file_.flush()
-            pje = PackageJSONExtractor(file_.name)
-            assert pje.data == data
+            mje = ManifestJSONExtractor(file_.name)
+            assert mje.data == data
```
Comment (magopian) [inline apps/files/tests/test_utils_.py]: The changes in this file aren't related to this bug, I just stumbled on them and saw they needed fixing. I can split them in another PR if needed.

**Comment by magopian** [comment]
@EnTeQuAk mind having a look at those three PRs that go together? #890 #889 and https://github.com/mozilla/amo-validator/pull/338 ?

**Comment by mstriemer** [comment]
Looks like the build is failing.

File: apps/files/tests/test_utils_.py (hunk 1 above)
Comment (mstriemer) [inline apps/files/tests/test_utils_.py]: Should these changes go in #889? Perhaps it makes sense to merge the two PRs but leave it as two separate commits?

**Comment by magopian** [comment]
Should be fixed now :)

File: apps/devhub/tests/test_views.py [hunk 2]
```diff
@@ -2222,6 +2222,30 @@ def test_unparsable_xpi(self, flag_is_active, v):
...
+    def test_experiment_xpi_allowed(self, mock_validator):
+        user = UserProfile.objects.get(email='regular@mozilla.com')
+        self.grant_permission(user, 'Experiments:submit')
+        mock_validator.return_value = json.dumps(self.validation_ok())
+        self.upload_file('../../../files/fixtures/files/experiment.xpi')
+        upload = FileUpload.objects.get()
+        response = self.client.get(reverse('devhub.upload_detail',
+                                           args=[upload.uuid, 'json']))
+        data = json.loads(response.content)
+        assert data['validation']['messages'] == []
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: Should you check that it was created in the database?

File: apps/devhub/tests/test_views.py [hunk 3]
```diff
@@ -2222,6 +2222,30 @@ def test_unparsable_xpi(self, flag_is_active, v):
...
+    def test_experiment_xpi_not_allowed(self, mock_validator):
+        mock_validator.return_value = json.dumps(self.validation_ok())
+        self.upload_file('../../../files/fixtures/files/experiment.xpi')
+        upload = FileUpload.objects.get()
+        response = self.client.get(reverse('devhub.upload_detail',
+                                           args=[upload.uuid, 'json']))
+        data = json.loads(response.content)
+        assert data['validation']['messages'] == [
+            {u'tier': 1, u'message': u'You cannot submit this type of add-ons',
+             u'fatal': True, u'type': u'error'}]
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: Likewise it might be good to check that it doesn't exist in the database but in this case I'm fairly confident it wouldn't be.

File: apps/devhub/tests/test_views.py [hunk 4]
```diff
@@ -2871,6 +2895,26 @@ def test_unlisted_addon_pass_validation(self, mock_sign_file):
...
 
+    @mock.patch('devhub.views.sign_file')
+    def test_experiments_are_auto_signed(self, mock_sign_file):
+        """Experiment extensions (bug 1220097) are auto-signed."""
+        self.upload = self.get_upload('experiment.xpi')
+        self.addon.update(guid='experiment@xpi', is_listed=True,
+                          status=amo.STATUS_PUBLIC, trusted=False)
+        # We're going to sign even if it has errors/warnings.
+        self.upload.update(
+            validation='{"notices": 2, "errors": 1, "messages": [],'
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: I'd suggest calling `json.dumps` instead of manually writing JSON but that isn't critical.

File: apps/devhub/tests/test_views.py [hunk 5]
```diff
@@ -2871,6 +2895,26 @@ def test_unlisted_addon_pass_validation(self, mock_sign_file):
...
+                       ' "metadata": {}, "warnings": 1,'
+                       ' "signing_summary": {"trivial": 1, "low": 0,'
+                       '                     "medium": 0, "high": 1},'
+                       ' "passed_auto_validation": 0}')
+        self.post()
+        assert self.addon.reload().status == amo.STATUS_PUBLIC
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: You set this to `amo.STATUS_PUBLIC` at the start but I can't think of why it would change.

File: apps/devhub/tests/test_views.py [hunk 6]
```diff
@@ -2871,6 +2895,26 @@ def test_unlisted_addon_pass_validation(self, mock_sign_file):
...
+        assert mock_sign_file.called
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: Do we know it was called with a file for this addon?

File: apps/devhub/tests/test_views.py [hunk 7]
```diff
@@ -2871,6 +2895,26 @@ def test_unlisted_addon_pass_validation(self, mock_sign_file):
         assert file_.status == amo.STATUS_LITE
         assert mock_sign_file.called
 
+    @mock.patch('devhub.views.sign_file')
+    def test_experiments_are_auto_signed(self, mock_sign_file):
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: I might be missing something but I don't see how why this file is accepted, the other test had to give the user permissions.

Frankly this test doesn't make me feel any more confident about the code but I could be missing something that's happening in the base class.

File: apps/devhub/views.py [hunk 8]
```diff
@@ -831,6 +831,12 @@ def json_upload_detail(request, upload, addon_slug=None):
     if result['validation']:
         try:
             pkg = parse_addon(upload, addon=addon)
+            if (pkg.get('is_experiment', False) and
+                    not acl.action_allowed(request, 'Experiments', 'submit')):
```
Comment (mstriemer) [inline apps/devhub/views.py]: Can you move this into `parse_addon`? This code path isn't hit in every case where you can create an add-on but `parse_addon` should always be run. If we land this then I have a feeling anyone will be able to create an experiment through the signing API.

**Comment by mstriemer** [comment]
Can you add a test to https://github.com/mozilla/olympia/blob/6966546abd344d8421020150455de7dc279f8443/apps/signing/tests/test_views.py#L62 to verify that the API doesn't allow experiments for user's without the permission? You should be able to point it to your new XPI and copy/paste one of the other tests.

Since I think it will be accepted I'm r- for now. Looks good otherwise.

File: apps/files/tests/test_utils_.py (hunk 1 above)
Comment (magopian) [inline apps/files/tests/test_utils_.py]: The modifications in this file are totally unrelated to experiments (it's related to PR #839) that landed already, for webextensions).I'm going to remove them, and commit them on master (they're only typos).

File: apps/devhub/tests/test_views.py (hunk 2 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: What was created in the database? Only the `FileUpload` gets created by the `self.upload_file` test helper, and we already know it's created since we get it on line 2231.

File: apps/devhub/tests/test_views.py (hunk 3 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: There is no "creation" of entries in the database in this test, apart from the `FileUpload`, which is created as soon as a file is uploaded.

File: apps/devhub/tests/test_views.py (hunk 4 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: Yep, I'll do that (it was a copy/paste from other tests doing the same thing, I'll change it there too).

File: apps/devhub/tests/test_views.py (hunk 5 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: Uploading a new version can change the status of the add-on automatically (using signals). We don't care about this change in this test thought, I'll remove it.

File: apps/devhub/tests/test_views.py (hunk 7 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: This isn't testing the same thing: when we upload a new file (on the frontend), there's an ajax call to the validator, which fills in the `FileUpload.validation` field. And to do that, we do need the permission (I'll try to move this permission checking code will to `parse_addon` as you advised below).

This test is testing something else: that once a file has been submitted, it'll be automatically signed (that's when you click on the "submit" button on the file upload page).

File: apps/devhub/tests/test_views.py (hunk 6 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: Yep, I'll add that, thanks ;)

File: apps/devhub/views.py (hunk 8 above)
Comment (magopian) [inline apps/devhub/views.py]: We don't always have access to the `request` in `parse_addon`, as it's used in quite a few places, and some of them are not in the request flow (eg we use it for parsing automatically fetched langpacks using crons).

Also, `parse_addon` isn't concerned with user permissions at all, which should be checked in the view I believe. I guess we could have `check_addon` taking a request, and it would check the user permission, and then in turn call the `parse_addon`? And then the API would use `check_addon` instead?

Experiments are a very specific use case: "special" extensions that are only submitted by Mozilla, and I think we'll have some more special extension types in the future. We might want to have something better/cleaner in place to cope with those, but one step at a time ;)

Anyway, I would push back on this comment, and ask your thoughts on doing the user permission check in the API view like it is here (in another PR? the same?). What do you think @mstriemer ?

**Comment by magopian** [comment]
@mstriemer ok I've updated the PR with some modifications to the API to also not accept submission of experiments unless authorized (and the tests ;)

Let me know what you think!

File: apps/devhub/tests/test_views.py (hunk 2 above)
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: Oh, yeah that makes sense. I just saw it making a request after and figured that must be creating something but I don't think we'd do that on a `GET` ;)

File: apps/devhub/tests/test_views.py [hunk 9]
```diff
@@ -2871,6 +2897,30 @@ def test_unlisted_addon_pass_validation(self, mock_sign_file):
...
+        """Experiment extensions (bug 1220097) are auto-signed."""
+        self.upload = self.get_upload('experiment.xpi')
+        self.addon.update(guid='experiment@xpi', is_listed=True,
+                          status=amo.STATUS_PUBLIC, trusted=False)
+        # We're going to sign even if it has errors/warnings.
+        self.upload.update(
+            validation=json.dumps({
```
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: It looks like you can pass `validation` to `self.get_upload`. It looks to me like even though this is setting the validation to have errors `self.upload.valid` might still be `True` since the upload was created [without validation errors](https://github.com/mozilla/olympia/blob/9a9231e0c79b99e0ac0e42e74d6b65a970c0176a/apps/files/tests/test_models.py#L52) and [valid will only get set to `True`](https://github.com/mozilla/olympia/blob/9a9231e0c79b99e0ac0e42e74d6b65a970c0176a/apps/files/models.py#L605) not `False` when `validation` changes.

File: apps/devhub/tests/test_views.py (hunk 7 above)
Comment (mstriemer) [inline apps/devhub/tests/test_views.py]: Okay, yeah that makes sense.

File: apps/files/utils.py [hunk 10]
```diff
@@ -208,6 +212,8 @@ def find_type(self):
         # https://developer.mozilla.org/en-US/Add-ons/Install_Manifests#type
         self.package_type = self.find('type')
         if self.package_type and self.package_type in self.TYPES:
+            # If it's an experiment, we need to store that for later.
+            self.is_experiment = (self.package_type == '128')
```
Comment (mstriemer) [inline apps/files/utils.py]: I don't think those parens are needed. Is that a flake8 thing or does it just look weird without?

Comment (mstriemer) [inline apps/files/utils.py]: Is there a constant you can use instead of a raw `'128'`?

File: apps/signing/tests/test_views.py [hunk 11]
```diff
@@ -124,6 +126,35 @@ def test_version_already_uploaded(self):
...
 
+    @mock.patch('devhub.views.auto_sign_version')
+    def test_version_added_is_experiment(self, sign_version):
+        self.grant_permission(self.user, 'Experiments:submit')
+        guid = 'experiment@xpi'
+        qs = Addon.unfiltered.filter(guid=guid)
+        assert not qs.exists()
+        response = self.put(
+            addon=guid, version='0.1',
+            filename=os.path.join('apps/files/fixtures/files/experiment.xpi'))
```
Comment (mstriemer) [inline apps/signing/tests/test_views.py]: I think you can leave out the `os.path.join` here, I don't think it's doing anything.

File: apps/devhub/tests/test_views.py (hunk 9 above)
Comment (magopian) [inline apps/devhub/tests/test_views.py]: Good catch, thanks!

File: apps/signing/tests/test_views.py [hunk 12]
```diff
@@ -124,6 +126,35 @@ def test_version_already_uploaded(self):
...
+        assert response.status_code == 201
+        assert qs.exists()
+        addon = qs.get()
+        assert addon.has_author(self.user)
+        assert not addon.is_listed
+        assert addon.status == amo.STATUS_LITE
```
Comment (mstriemer) [inline apps/signing/tests/test_views.py]: This is failing on Travis. It normally gets set in the [`submit_file` task](https://github.com/mozilla/olympia/blob/9a9231e0c79b99e0ac0e42e74d6b65a970c0176a/apps/devhub/tasks.py#L83) but it probably isn't because `FileUpload.passed_all_validations` isn't `True`.

**Comment by mstriemer** [comment]
This is looking good. Thanks for the clarification in those tests, not sure why I couldn't figure them out before.

Are there any updates needed to the frontend to enable the submit button or are we assuming they'll have the "skip failed validation" option since we gave them the permission to upload experiments. And I guess once the Addon is created they can upload a new version to it without any issues?

File: apps/signing/tests/test_views.py (hunk 11 above)
Comment (magopian) [inline apps/signing/tests/test_views.py]: HAHA. Would be funny if it wasn't so sad...

File: apps/signing/tests/test_views.py (hunk 12 above)
Comment (magopian) [inline apps/signing/tests/test_views.py]: Yep, I had a fix already (hence my questions on irc ;), it should now be fixed

**Comment by magopian** [comment]
The udpates to the front-end is really unclear for me at the moment. I don't know if we're expecting the experiments to fail, have signing related warnings, if they're supposed to be unlisted or not, asking for full or prelim review...

I'd like to have this on -dev or stage asap, so it can be tested for real, and then we'll see what we need.

I've modified a few things, let me know how this looks ;)

**Comment by mstriemer** [comment]
I'm assuming people will want to use the API so it might not matter even.

**Comment by magopian** [comment]
Ok, this should not be merged "as is", I changed the requirements to use https://github.com/mozilla/amo-validator/pull/338

This should fix the failing test. If it does, then we'll be able to merge all three PRs (https://github.com/mozilla/amo-validator/pull/338, #889  and this one).

File: requirements/prod.txt [hunk 13]
```diff
@@ -1,5 +1,6 @@
 -r compiled.txt
-amo-validator==1.9.7
+#amo-validator==1.9.7
+-e git+https://github.com/magopian/amo-validator@e425576b948f5f967ff7fe00f687fe103b662b21#egg=amo-validator
```
Comment (EnTeQuAk) [inline requirements/prod.txt]: I gather you're removing this one the other PRs are merged right?

Comment (magopian) [inline requirements/prod.txt]: Yep! As soon as the other PR on the validator is merged, I can release a new version of the validator, and use it here instead.
